# EurasianOtter_TrophyPaws

Bioinformatic scripts for analyses of two historical Eurasian otter trophy hunting paws in the context of modern British variation. 
First author email <sarah.jane.du.plessis@gmail.com>.

### Citation:
Sarah J du Plessis, Nuno F Martins, Xin Sun, Sarah S T Mak, David W G Stanton, Klaus-Peter Koepfli, Elizabeth A Chadwick, M Thomas P Gilbert, Frank Hailer, Historical genomic variation of Eurasian otters (*Lutra lutra*) in Britain, from hunting trophies. *Conservation Genet Resour* (2025). [https://doi.org/10.1007/s12686-025-01383-9](https://doi.org/10.1007/s12686-025-01383-9)

### Abstract:
Generating genomic data from historical specimens can provide insight into populations and processes no longer evident from modern samples. Identifying viable sources of historical DNA is key to this. Here we aimed to explore the viability of trophy hunting paws as sources of historical DNA, and to use this data to evaluate changes in population structuring following a significant bottleneck. Further, we aimed to search for the presence of genomic signals of the putative recent introduction of Asian otters to the UK previously identified using modern data. Substantial proportions of endogenous DNA (77-85 %) were identified in both samples, illustrating the potential of such specimens for historical DNA research. Autosomal variation suggested that changes in population structure in Britain have occurred since the population bottleneck. Both mitochondrial and autosomal variation show no signs of the divergent genetic lineage identified in modern data, for example of the substantial private variation in modern data from the east of England, minimal proportions were identified in the historical specimens. This preliminary dataset indicates the huge potential of historical trophy paw samples to assess past population genetic dynamics, calling for further research in this area.

### Raw data:
Newly generated raw reads were deposited on SRA (BioProject PRJNA1081818, BioSamples SAMN40188547-8, SRA SRR28150401-2), and mitochondrial haplotypes on GenBank (PP409971-2). Previously published (du Plessis et al. 2023a) raw data were available through the DToL Data Portal ([https://portal.darwintreeoflife.org/data/root/details/Lutra%20lutra](https://portal.darwintreeoflife.org/data/root/details/Lutra%20lutra)) and haplotype data from NCBI Nucleotide Database (OR633269-86 and BK064833-5). Bioinformatic code used to conduct all analyses available on GitHub (https://github.com/sduplessis1/EurasianOtter_TrophyPaws) and in electronic supplementary material.

du Plessis, S. J., Blaxter, M., Koepfli, K.-P., Chadwick, E. A. and Hailer, F. 2023a. Genomics reveals complex population history and unexpected diversity of Eurasian otters (*Lutra lutra*) in Britain relative to genetic methods. *Molecular Biology and Evolution*, p. msad207. [https://doi.org/10.1093/molbev/msad207](https://doi.org/10.1093/molbev/msad207)
